spring-cloud-app-broker-logging/src/main/java/org/springframework/cloud/appbroker/logging/streaming/LogCacheStreamPublisher.java

```diff
@@ -0,0 +1,120 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.util.retry.Retry;
+
+import org.springframework.cloud.appbroker.logging.ApplicationIdsProvider;
+import org.springframework.cloud.appbroker.logging.LoggingUtils;
+
+public class LogCacheStreamPublisher implements LogStreamPublisher<Envelope> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(LogCacheStreamPublisher.class);
+
+	private final CloudFoundryClient client;
+
+	private final LogCacheClient logCacheClient;
+
+	private final ApplicationIdsProvider applicationIdsProvider;
+
+	public LogCacheStreamPublisher(
+		CloudFoundryClient client,
+		LogCacheClient logCacheClient,
+		ApplicationIdsProvider applicationIdsProvider) {
+		this.client = client;
+		this.logCacheClient = logCacheClient;
+		this.applicationIdsProvider = applicationIdsProvider;
+	}
+
+	@Override
+	public Flux<Envelope> getLogStream(String serviceInstanceId) {
+		return this.applicationIdsProvider
+			.getApplicationIds(serviceInstanceId)
+			.doOnNext(id -> LOG.debug("Starting log streaming for app with ID {}", id))
+			.flatMap(this::createApplicationStreamer);
+	}
+
+	protected Flux<Envelope> createApplicationStreamer(String applicationId) {
+		return client.applicationsV2()
+			.get(GetApplicationRequest.builder()
+				.applicationId(applicationId)
+				.build())
+			.map(response -> response.getEntity().getName())
+			.flatMapMany(appName -> {
+				long initialStartTime = Instant.now().minus(5, ChronoUnit.SECONDS).toEpochMilli() * 1_000_000L;
+				return logCacheClient.read(
+						ReadRequest.builder()
+							.sourceId(applicationId)
+							.envelopeTypes(EnvelopeType.LOG)
+							.startTime(initialStartTime)
+							.build())
+					.flatMapMany(initialResponse -> {
+						AtomicLong lastTimestamp = new AtomicLong(
+							initialResponse.getEnvelopes().getBatch().stream()
+								.mapToLong(org.cloudfoundry.logcache.v1.Envelope::getTimestamp)
+								.max()
+								.orElse(initialStartTime)
+						);
+
+						Flux<Envelope> initialLogs = Flux.fromIterable(initialResponse.getEnvelopes().getBatch())
+							.map(LoggingUtils::convertLogCacheEnvelopeToDropsonde);
+
+						Flux<Envelope> polledLogs = Flux.interval(Duration.ofSeconds(1))
+							.flatMap(tick -> logCacheClient.read(
+									ReadRequest.builder()
+										.sourceId(applicationId)
+										.envelopeTypes(EnvelopeType.LOG)
+										.startTime(lastTimestamp.get() + 1)
+										.build())
+								.flatMapMany(readResponse -> {
+									long maxTimestamp = readResponse.getEnvelopes().getBatch().stream()
+										.mapToLong(org.cloudfoundry.logcache.v1.Envelope::getTimestamp)
+										.max()
+										.orElse(lastTimestamp.get());
+
+									lastTimestamp.set(maxTimestamp);
+
+									return Flux.fromIterable(readResponse.getEnvelopes().getBatch())
+										.map(LoggingUtils::convertLogCacheEnvelopeToDropsonde);
+								}))
```

Comment:
This is pretty difficult to follow. Could some of these lambdas be extracted into well-named methods?

Reply:
You are right. Extracted a few functions, thanks!